Rfc/unwind auth (#2367)
neo4j/graphql: docs/rfcs/rfc-029-unwind-create-auth.md

    ```graphql
    type User {
        id: ID @auth(rules: [{ operations: [CREATE], roles: ["admin"] }])
        name: String
    }

    type Post {
        title: String
        content: String
        moderators: [User!]! @relationship(type: "MODERATES_POST", direction: IN)
    }
    ```

-   ### Particular case: Entity-type Bind - across relationship

    **Schema**

    ```graphql
    type User {
        id: ID
        name: String
    }
    type Post @auth(rules: [{ operations: [CREATE], bind: { creator: { id: "$jwt.sub" } } }]) {
        title: String
        content: String
        creator: [User!]! @relationship(type: "HAS_POST", direction: IN)
    }
    ```

### Entity-type solution

Entity type `@auth` as well as Nested Entity type `@auth` no needs particular consideration when used in the unwind-create context.

It could be appended directly on the entity modified,
For instance, preventing a post is created by users without the `ADMIN` role, could be achieved as follow:

```cypher
UNWIND $create_param0 AS create_var1
CALL {
    WITH create_var1
    CREATE (create_this0:`Post`)
    SET
        create_this0.content = create_var1.content
    WITH create_this0, create_var1

>     CALL apoc.util.validate(NOT (any(auth_var1 IN ["admin"] WHERE any(auth_var0 IN $auth.roles WHERE auth_var0 = auth_var1))), "@neo4j/graphql/FORBIDDEN", [0])

Reviewer: I would consider using `validatePredicate` instead here - I think it's nicer than having procedure calls spattered about! Consider this comment to be applied across this PR.

```suggestion
    WHERE apoc.util.validatePredicate(NOT (any(auth_var1 IN ["admin"] WHERE any(auth_var0 IN $auth.roles WHERE auth_var0 = auth_var1))), "@neo4j/graphql/FORBIDDEN", [0])
```

            "name": "Simone"
          }
        }
      }
    }
  ] AS create_var2
CALL {
    WITH create_var2
    CREATE (create_this1:`Post`)
    SET
        create_this1.title = create_var2.title,
        create_this1.content = create_var2.content
    WITH create_this1, create_var2
    CALL {
        WITH create_this1, create_var2
        UNWIND create_var2.creator.create AS create_var3
        WITH create_var3.node AS create_var4, create_var3.edge AS create_var5, create_this1
        CREATE (create_this6:`User`)
        SET
            create_this6.id = create_var4.id,
            create_this6.name = create_var4.name
        MERGE (create_this6)-[create_this7:HAS_POST]->(create_this1)
        
        RETURN collect(NULL)
    }
    WITH create_this1
    CALL {
        WITH create_this1
        MATCH (create_this1)<-[create_this1_creator_User_unique:HAS_POST]-(:User)
        WITH count(create_this1_creator_User_unique) as c
        CALL apoc.util.validate(NOT (c = 1), '@neo4j/graphql/RELATIONSHIP-REQUIREDPost.creator required', [0])
        RETURN c AS create_this1_creator_User_unique_ignored
    }
    WITH create_this1
    CALL {
            WITH create_this1

>             CALL apoc.util.validate(
>                 NOT ((exists((create_this1)<-[:HAS_POST]-(:`User`)) 
>                     AND 
>             all(auth_this0 IN [(create_this1)<-[:HAS_POST]-(auth_this0:`User`) | auth_this0]
>                  WHERE (auth_this0.id IS NOT NULL AND auth_this0.id = $this0auth_param0))))
>                  , "@neo4j/graphql/FORBIDDEN", [0])

Reviewer: This whole Cypher needs a whole lot of modernising, but unsure if that's outside of the scope of this work:

```suggestion
            WHERE apoc.util.validatePredicate(
                NOT EXISTS { (create_this1)<-[:HAS_POST]-(:`User`) }
                    AND 
                NOT EXISTS { (create_this1)<-[:HAS_POST]-(auth_this0:`User`) 
                 WHERE NOT auth_this0.id IS NOT NULL AND auth_this0.id = $this0auth_param0 }
                 , "@neo4j/graphql/FORBIDDEN", [0])
```

            RETURN create_this1
    }
    RETURN create_this1
}
CALL {
    WITH create_this1
    MATCH (create_this1_creator:`User`)-[create_this0:HAS_POST]->(create_this1)
    WITH create_this1_creator { .id, .name } AS create_this1_creator
    RETURN head(collect(create_this1_creator)) AS create_this1_creator
}
RETURN collect(create_this1 { .title, .content, creator: create_this1_creator }) AS data
```

## Out of scope

-   Subscription rules.
-   `Allow` rules.
-   Rules on `connect` or `connectOrCreate` operations.

Full examples:

> ## Roles rule

Reviewer: I'm not too fussed about roles - get it working with as little work as possible, because we will soon resolve this at the JavaScript level rather than in the database.
